tensorflow/core/kernels/adjust_hue_op_gpu.cu.cc

```diff
@@ -0,0 +1,182 @@
+limitations under the License.
+==============================================================================*/
+#if GOOGLE_CUDA
+
+#define EIGEN_USE_GPU
+
+#include "adjust_hue_op.h"
+#include "tensorflow/core/util/cuda_kernel_helper.h"
+
+namespace tensorflow {
+
+typedef Eigen::GpuDevice GPUDevice;
+
+namespace internal {
+
+typedef struct RgbTuple {
+
+  float r, g, b;
+
+} RgbTuple;
+
+typedef struct HsvTuple {
+
+  float h, s, v;
+
+} HsvTuple;
+
+__device__ HsvTuple rgb2hsv_cuda(const float r, const float g, const float b) {
+
+  HsvTuple tuple;
+
+  const float M = fmaxf(r, fmaxf(g, b));
+  const float m = fminf(r, fminf(g, b));
+  const float chroma = M - m;
+
+  float h = 0.0f, s = 0.0f;
+
+    // hue
+    if (chroma > 0.0f) {
+      if (M == r) {
+
+        const float num = (g - b) / chroma;
+        const float sgn = num < 0.0f;
+        const float sign = powf(-1.0f, sgn);
+        h = (sgn * 6.0f + sign * fmodf(sign * num, 6.0f)) / 6.0f;
+
+      } else if (M == g) {
+
+        h = ((b - r) / chroma + 2.0) / 6.0f;
+
+      } else {
+
+        h = ((r - g) / chroma + 4.0) / 6.0f;
+      }
+
+    } else {
+
+      h = 0.0f;
+    }
+
+   // saturation
+   if (M > 0.0) {
+
+     s = chroma / M;
+
+   } else {
+
+     s = 0.0f;
+   }
+
+  tuple.h = h;
+  tuple.s = s;
+  tuple.v = M;
+
+  return tuple;
+}
+
+__device__ RgbTuple hsv2rgb_cuda(const float h, const float s, const float v) {
+
+  RgbTuple tuple;
+
+  const float new_h = h * 6.0f;
+  const float chroma = v * s;
+  const float x = chroma * (1.0 - fabsf(fmodf(new_h, 2.0f) - 1.0f));
+
+  const float new_m = v - chroma;
+  const bool between_0_and_1 = new_h >= 0.0 && new_h < 1;
+  const bool between_1_and_2 = new_h >= 1.0 && new_h < 2;
+  const bool between_2_and_3 = new_h >= 2 && new_h < 3;
+  const bool between_3_and_4 = new_h >= 3 && new_h < 4;
+  const bool between_4_and_5 = new_h >= 4 && new_h < 5;
+  const bool between_5_and_6 = new_h >= 5 && new_h < 6;
+
+  tuple.r = chroma * (between_0_and_1 || between_5_and_6) +
+                  x * (between_1_and_2 || between_4_and_5) + new_m;
+
+  tuple.g = chroma * (between_1_and_2 || between_2_and_3) +
+                      x * (between_0_and_1 || between_3_and_4) + new_m;
+
+  tuple.b =  chroma * (between_3_and_4 || between_4_and_5) +
+                      x * (between_2_and_3 || between_5_and_6) + new_m;
+
+  return tuple;
+}
+
+__global__ void adjust_hue_nhwc(const int number_elements,
+                                const float * const input, float * const output, const float * const hue_delta) {
+
+        const float delta = hue_delta[0];
+
+	// multiply by 3 since we're dealing with contiguous RGB bytes for each pixel
+	const int idx = (blockDim.x * blockIdx.x + threadIdx.x) * 3;
+
+	// bounds check
+	if (idx > number_elements) {
+		return;
+	}
+
+        const HsvTuple hsv = rgb2hsv_cuda(input[idx], input[idx+1], input[idx+2]);
+
+	// hue adjustment
+	const float new_h = fmodf(hsv.h + delta, 1.0f);
```

Comment:
@zheng-xq It does, plus since the fmodf is around 1.0f, the delta can be even a multiple of 1 or -1. But at a minimum, it does support the range [-1, 1].